File: COSC 4315/assign2/Scala/test/TailRecursionTest.scala
import org.junit.Test
import org.junit.Assert._
import org.scalatest._

class TailRecursionTest extends FlatSpec{

    val factorial = new TailRecursiveFact()

    val fiftyFactorial = BigInt("30414093201713378043612608166064768844377641568960512000000000000")

    @Test
    def canaryTest(){
        assertTrue(true)
    }

    @Test
    def outOfRangeTest(){
        assertEquals(BigInt(-1), factorial.tailRecursiveFact(BigInt(-5)))
    }

    @Test
    def zeroFactorialTest(){
        assertEquals(BigInt(1), factorial.tailRecursiveFact(BigInt(0)))
    }

    @Test
    def factorialTest01(){
        assertEquals(BigInt(1), factorial.tailRecursiveFact(BigInt(1)))
    }

    @Test
    def factorialTest05(){
        assertEquals(BigInt(120), factorial.tailRecursiveFact(BigInt(5)))
    }

    @Test
    def factorialTest10(){
        assertEquals(BigInt(3628800), factorial.tailRecursiveFact(BigInt(10)))
    }

    @Test
    def factorialTest50(){
        assertEquals(fiftyFactorial, factorial.tailRecursiveFact(BigInt(50)))
    }

    @Test
    def testTailRecursionIsTCO(){
        assertNotNull(factorial.tailRecursiveFact(BigInt(10000)))
    }
}

File: COSC 4315/assign2/Kotlin/test/TailRecursionTest.kt
import org.junit.Test
import org.junit.Assert.*
import java.math.BigInteger

class TailRecursionTest {

    val tailFactorial = TailRecursionFact()

    val fiftyFactorial = BigInteger("30414093201713378043612608166064768844377641568960512000000000000")

    @Test
    fun canaryTest(){
        assertTrue(true)
    }

    @Test
    fun outOfRangeTest(){
        assertEquals(BigInteger.valueOf(-1), tailFactorial.tailRecursionFact(-5))
    }

    @Test
    fun zeroFactorialTest(){
        assertEquals(BigInteger.valueOf(1), tailFactorial.tailRecursionFact(0))
    }

    @Test
    fun factorialTest01(){
        assertEquals(BigInteger.valueOf(1), tailFactorial.tailRecursionFact(1))
    }

    @Test
    fun factorialTest02(){
        assertEquals(BigInteger.valueOf(120), tailFactorial.tailRecursionFact(5))
    }

    @Test
    fun factorialTest03(){
        assertEquals(BigInteger.valueOf(3628800), tailFactorial.tailRecursionFact(10))
    }

    @Test
    fun factorialTest04(){
        assertEquals(fiftyFactorial, tailFactorial.tailRecursionFact(50))
    }

    @Test
    fun isTailCallOptimized(){
        assertNotNull(tailFactorial.tailRecursionFact(10000))
    }
}